service/history/chasm_notifier.go

```diff
@@ -0,0 +1,46 @@
+package history
+
+import (
+	"sync"
+
+	"go.temporal.io/server/chasm"
+	"go.temporal.io/server/common/metrics"
+)
+
+type (
+	// ChasmNotifier allows subscribers to receive notifications relating to a CHASM execution.
+	ChasmNotifier struct {
+		executions map[chasm.EntityKey]chan struct{}
+		lock       sync.Mutex
+	}
+)
+
+// NewChasmNotifier creates a new instance of ChasmNotifier.
+func NewChasmNotifier(metricsHandler metrics.Handler) *ChasmNotifier {
+	return &ChasmNotifier{
+		executions: make(map[chasm.EntityKey]chan struct{}),
+	}
+}
+
+// Subscribe returns a channel that will be closed when there is a notification relating to the
+// execution. No data will be written to the channel.
+func (n *ChasmNotifier) Subscribe(key chasm.EntityKey) <-chan struct{} {
```

Comment:
You are going to want an `Unsubscribe` function too in case the notification never comes in.
For that you're going to need a waiter count per key to know if you can delete the key from the map and you should only delete if the channel hasn't been updated in the map.

```go
type subscriptionTracker struct {
	ch chan struct{}
	numWaiters int
}
```


```suggestion
func (n *ChasmNotifier) Subscribe(key chasm.EntityKey) (<-chan struct{}, func()) {
```